checkqueue: implement a new scriptcheck worker pool with atomic variables

HowHsu opened the pull request:
TL;DR
    This patchset proposes a new design of scriptcheck worker pool with
    atomic variables to leverage more cpu resource on a machine which the
    current mutex version cannot. Achieve about 45% boost in performance.
    
    Main Idea:
    
    The current version of worker pool use mutex to do the synchronization
    between worker threads, which makes the performance downgrade after
    worker_threads_num >= 14. One root cause is the high contention on the
    queue's mutex among worker threads. Workers repeatedly sleep and be woken
    up in short time, leading to heavy context switching.
    The idea is to use atomic variables to protect the queue, workers
    acquire it and increase it with a batch_size(a small number). This
    significantly reduces context switching.
    
    Implementation
    
    To implement it, below changes are made:(enabled in atomic mode)
    - atomic<int> nNext is used to indicate the next available indice in the
      queue
    - atomic<int> nTodo2 is used to indicate the number of tasks left
    - nBatchSize is the batch size of tasks everytime a worker consumes
      it should be small to keep workload balance among workers
    - m_mutex is still used, to protect m_result and conditional vars
    - fAtomic to indicate which mode is enabled
    - about the queue:
    
      - the queue is used as queue, not stacks any more
      - the queue is 'add only', which means only adding tasks to it, don't
        remove used tasks.
    
        the above two are a must to make this idea achievable, and the
        second sacrifices space
    
      - start the workers only after all the tasks have been added to the
        queue
    
        this is a trade-off to make the implementaion simple, a worker can
        compare nNext + nBatchSize to queue.size() to find if there are
        tasks undone in the queue, otherwise another atomic<int> tail is
        needed, and race condition may exist in this way (need more
        investigation). From my test, it's not 100% to say this is a
        'trade-off' since it also brings something good, pros && cons:
        - pros:
            1. no contention among producer and consumers when adding tasks
               to the queue
            2. the adding period is very short compared to the verification
               period
        - cons:
            1. the workers start later
    
    For details, please refer to the worker handler code
    
    Tests:
    
    - Numbers
      - For blocks/s numbers
        `./build/bin/bench_bitcoin --filter=ConnectBlockMixedEcdsaSchnor`
      - For cpu usage numbers
        `NANOBENCH_ENDLESS=ConnectBlockAllEcdsa ./build/bin/bench_bitcoin --filter=ConnectBlockAllEcdsa`
    
      - Some data is missing because the machine was billed by time, and there wasn't
        enough time to complete all measurements.
      - Some cells contain a range instead of a single value due to high fluctuations
      - The numbers represent the average of multiple test runs.
    
               |     mutex verion       |   atomic version
    nr_workers | blocks/s  | cpu usage  | blocks/s  | cpu usage2
    -----------|-----------|------------|-----------|-------------
    0          | 4.22      | 100%       | 4.2       | 100%
    1          | 8.23      | 198%       | 8         | 191.3%
    2          | 12.08     | 293.7%     | 11.6      | 276%
    3          | 16        | 386.7%     | 14.9      | 354.5%
    4          | 19.8      | 476.3%     |           |
    5          | 23.12     | 565.4%     |           |
    6          | 27.3      | 654%       | 23        | 560%
    7          | 28.5      | 738%       |           |
    8          | 33.8      | 809%       |           |
    9          | 33.5      | 868%       | 28.5      | 730%
    10         | 37.4      | 910%       | 30.5      | 777%
    11         | 33~40     | 880%~928%  | 32        | 825%
    12         | 22~39     | 860%~925%  | 32.5      | 830%
    13         | 27.8~37   | 700%~913%  | 34.8      | 928%
    15         | 24.8~38.9 | 640%~719%  | 38        | 1000%
    19         | 18~23     | 370%~510%  | 41        | 1161%
    22         | 7~18      | 270%~400%  |           |
    25         | 6.4       | 260%       | 46        | 1400%
    30         | 5         | 270%       | 49        | 1560%
    60         | 5         | 280%       | 50~58     | 2100%
    
    - Env:
     A virtual machine on cloud
       Architecture:          x86_64
       CPU op-mode(s):        32-bit, 64-bit
       Address sizes:         46 bits physical, 48 bits virtual
       Byte Order:            Little Endian
       CPU(s):                64
       On-line CPU(s) list:   0-63
       Vendor ID:             GenuineIntel
       Model name:            Intel Xeon Processor (Cascadelake)
       CPU family:            6
       Model:                 85
       Thread(s) per core:    2
       Core(s) per socket:    16
       Socket(s):             2
       Stepping:              6
       BogoMIPS:              5985.93
     Virtualization features:
       Hypervisor vendor:     KVM
       Virtualization type:   full
     Caches (sum of all):
       L1d:                   2 MiB (64 instances)
       L1i:                   2 MiB (64 instances)
       L2:                    128 MiB (32 instances)
       L3:                    32 MiB (2 instances)
     NUMA:
       NUMA node(s):          2
       NUMA node0 CPU(s):     0-31
       NUMA node1 CPU(s):     32-63
    
    Signed-off-by: Hao Xu <hao.xu@linux.dev>

romanz committed:
rest: fetch spent transaction outputs by blockhash

Today, it is possible to fetch a block's spent prevouts in order to
build an external index by using the `/rest/block/HASH.json` endpoint.
However, its performance is low due to JSON serialization overhead.

We can significantly optimize it by adding a new REST endpoint, using
a binary response format:

```
$ BLOCKHASH=00000000000000000002a7c4c1e48d76c5a37902165a270156b7a8d72728a054

$ ab -k -c 1 -n 100 http://localhost:8332/rest/block/$BLOCKHASH.json
Document Length:        13278152 bytes
Requests per second:    3.53 [#/sec] (mean)
Time per request:       283.569 [ms] (mean)

$ ab -k -c 1 -n 10000 http://localhost:8332/rest/spentoutputs/$BLOCKHASH.bin
Document Length:        195591 bytes
Requests per second:    254.47 [#/sec] (mean)
Time per request:       3.930 [ms] (mean)
```

Currently, this PR is being used and tested by Bindex:

 * https://github.com/romanz/bindex-rs

This PR would allow to improve the performance of external indexers
such as electrs, ElectrumX, Fulcrum and Blockbook:

 * https://github.com/romanz/electrs (also https://github.com/Blockstream/electrs and https://github.com/mempool/electrs)
 * https://github.com/spesmilo/electrumx
 * https://github.com/cculianu/Fulcrum
 * https://github.com/trezor/blockbook

romanz committed:
doc: add release notes for #32540

HowHsu committed:
checkqueue: implement a new scriptcheck worker pool with atomic variables

TL;DR
This patchset proposes a new design of scriptcheck worker pool with
atomic variables to leverage more cpu resource on a machine which the
current mutex version cannot. Achieve about 45% boost in performance.

Main Idea:

The current version of worker pool use mutex to do the synchronization
between worker threads, which makes the performance downgrade after
worker_threads_num >= 14. One root cause is the high contention on the
queue's mutex among worker threads. Workers repeatedly sleep and be woken
up in short time, leading to heavy context switching.
The idea is to use atomic variables to protect the queue, workers
acquire it and increase it with a batch_size(a small number). This
significantly reduces context switching.

Implementation

To implement it, below changes are made:(enabled in atomic mode)
- atomic<int> nNext is used to indicate the next available indice in the
  queue
- atomic<int> nTodo2 is used to indicate the number of tasks left
- nBatchSize is the batch size of tasks everytime a worker consumes
  it should be small to keep workload balance among workers
- m_mutex is still used, to protect m_result and conditional vars
- fAtomic to indicate which mode is enabled
- about the queue:

  - the queue is used as queue, not stacks any more
  - the queue is 'add only', which means only adding tasks to it, don't
    remove used tasks.

    the above two are a must to make this idea achievable, and the
    second sacrifices space

  - start the workers only after all the tasks have been added to the
    queue

    this is a trade-off to make the implementaion simple, a worker can
    compare nNext + nBatchSize to queue.size() to find if there are
    tasks undone in the queue, otherwise another atomic<int> tail is
    needed, and race condition may exist in this way (need more
    investigation). From my test, it's not 100% to say this is a
    'trade-off' since it also brings something good, pros && cons:
    - pros:
        1. no contention among producer and consumers when adding tasks
           to the queue
        2. the adding period is very short compared to the verification
           period
    - cons:
        1. the workers start later

For details, please refer to the worker handler code

Tests:

- Numbers
  - For blocks/s numbers
    `./build/bin/bench_bitcoin --filter=ConnectBlockMixedEcdsaSchnor`
  - For cpu usage numbers
    `NANOBENCH_ENDLESS=ConnectBlockAllEcdsa ./build/bin/bench_bitcoin --filter=ConnectBlockAllEcdsa`

  - Some data is missing because the machine was billed by time, and there wasn't
    enough time to complete all measurements.
  - Some cells contain a range instead of a single value due to high fluctuations
  - The numbers represent the average of multiple test runs.

           |     mutex verion       |   atomic version
nr_workers | blocks/s  | cpu usage  | blocks/s  | cpu usage2
-----------|-----------|------------|-----------|-------------
0          | 4.22      | 100%       | 4.2       | 100%
1          | 8.23      | 198%       | 8         | 191.3%
2          | 12.08     | 293.7%     | 11.6      | 276%
3          | 16        | 386.7%     | 14.9      | 354.5%
4          | 19.8      | 476.3%     |           |
5          | 23.12     | 565.4%     |           |
6          | 27.3      | 654%       | 23        | 560%
7          | 28.5      | 738%       |           |
8          | 33.8      | 809%       |           |
9          | 33.5      | 868%       | 28.5      | 730%
10         | 37.4      | 910%       | 30.5      | 777%
11         | 33~40     | 880%~928%  | 32        | 825%
12         | 22~39     | 860%~925%  | 32.5      | 830%
13         | 27.8~37   | 700%~913%  | 34.8      | 928%
15         | 24.8~38.9 | 640%~719%  | 38        | 1000%
19         | 18~23     | 370%~510%  | 41        | 1161%
22         | 7~18      | 270%~400%  |           |
25         | 6.4       | 260%       | 46        | 1400%
30         | 5         | 270%       | 49        | 1560%
60         | 5         | 280%       | 50~58     | 2100%

- Env:
 A virtual machine on cloud
   Architecture:          x86_64
   CPU op-mode(s):        32-bit, 64-bit
   Address sizes:         46 bits physical, 48 bits virtual
   Byte Order:            Little Endian
   CPU(s):                64
   On-line CPU(s) list:   0-63
   Vendor ID:             GenuineIntel
   Model name:            Intel Xeon Processor (Cascadelake)
   CPU family:            6
   Model:                 85
   Thread(s) per core:    2
   Core(s) per socket:    16
   Socket(s):             2
   Stepping:              6
   BogoMIPS:              5985.93
 Virtualization features:
   Hypervisor vendor:     KVM
   Virtualization type:   full
 Caches (sum of all):
   L1d:                   2 MiB (64 instances)
   L1i:                   2 MiB (64 instances)
   L2:                    128 MiB (32 instances)
   L3:                    32 MiB (2 instances)
 NUMA:
   NUMA node(s):          2
   NUMA node0 CPU(s):     0-31
   NUMA node1 CPU(s):     32-63

Signed-off-by: Hao Xu <hao.xu@linux.dev>

HowHsu committed:
checkqueue: set max number of scriptcheck threads to GetNumCores()

The current max number of scriptcheck threads is MAX_SCRIPTCHECK_THREADS = 14
due to the limitation of the current mechanism, release it to GetNumCores() as
we have atomic mode of scriptcheck worker pool.

Signed-off-by: Hao Xu <hao.xu@linux.dev>

HowHsu committed:
checkqueue: fix: add common/system.cpp to libbitcoinkernel

GetNumCores() is now used in checkqueue related code, thus it has to be
added to libbitcoinkernel to keep everything ok

Signed-off-by: Hao Xu <hao.xu@linux.dev>

Muaytie666 reviewed (commented):
98b1db9283b5c04775309fa5dc2161fc597b355a

HowHsu commented:
hmmm...Ok, I realized that this may only be suitable for when all the preout Coin are in CCoinsViewCache, otherwise the `Add()` is not fast due to I/O there.

HowHsu commented:
> Hint: Call /ci_container_base/test/functional/combine_logs.py '/ci_container_base/ci/scratch/test_runner/test_runner_₿_🏃_20250626_150635/wallet_fundrawtransaction_252'

In my local run, I can do something like this to see the bitcoind log (node log as I said), but how can I see the bitcoind log on Cirrus CI, seems no way to do that since that is not exposed.

HowHsu commented:
> > is not exposed.
> 
> it is: See " View more details on Cirrus CI " on the checks page

Sorry, there is node logs, I missed them, thanks again.

HowHsu committed:
tmp: this is to test the CI failure

Signed-off-by: Hao Xu <hao.xu@linux.dev>

HowHsu committed:
checkqueue: mark and quit fast when error result is raised

Use a atomic<bool> fResultDone to indicate the result is out to be
something error, thus all workers can quit immediately. This is to
boost the performance in script verification error case.

Signed-off-by: Hao Xu <hao.xu@linux.dev>

instagibbs commented:
Are the first two commits related to the PR? Seem REST-related?

HowHsu commented:
> > hmmm...Ok, I realized that this may only be suitable for when all the preout Coin are in CCoinsViewCache, otherwise the `Add()` is not fast due to I/O there.
> 
> With #31132 , this one makes sense again, need to test them together in all-cache-miss case

Tested this one and [](https://github.com/bitcoin/bitcoin/pull/31132) combined with about 100+ contiguous real blocks in mainnet from height 840000, by the method I mentioned in [](https://github.com/bitcoin/bitcoin/issues/33375#issuecomment-3292198687).
But the result tells another story, this atomic version of work pool doesn't perform better than the master branch one, while it's a result on a 15-physical-core machine, I'll test it on a machine with more cores when I manage to get such a machine..

darosior reviewed (commented):
The current design is a queue protected by a mutex. The work producer locks the mutex when adding new checks to the queue, and the work consumers lock the mutex when moving checks out of the queue.

What you propose here is to avoid the mutex on the queue entirely, by having consumers simply read work from the queue and update an atomic index representing the last read job. This reduces lock contention since pushing work on the queue is now non-blocking, and reading work does not involve taking a whole mutex but simply reading-and-updating an atomic. Am i understanding what you suggest correctly?

If so, how is this safe at all? The producer pushing new work to the unprotected queue may lead to a vector reallocation while the consumer threads are still reading it. In your implementation the consumers also read the vector size to get the number of jobs available, while the producer writes to this vector concurrently. As far as i know `std::vector` gives you no guarantee that this is safe to do.

HowHsu commented:
> Also, could you cleanly rebase your branch on master?

Hi @darosior , I may return to this PR weeks later, busy with other PR stuff now, sorry for the delay.

maflcko commented:
are you still working on this, or can it be closed?

HowHsu commented:
Close this for now, I may reopen this later when I have time on it.
